### PR TITLE
minimal repro for fsdp + tp incorrect permutation

### PR DESCRIPTION
**Pull request by wesleytruong**
This PR minimally reproduces the incorrectness when loading from a checkpoint in hf with tp and fsdp enabled. The train script immediately saves after loading so that we can test the dcp checkpoint for correctness without training.

To reproduce the fsdp+tp incorrectness:
1. Load from hf checkpoint and create dcp checkpoint for testing:
`NGPU=4 CONFIG_FILE=./torchtitan/models/llama3/train_configs/llama3_8b_debug.toml ./run_train.sh`
2. Sanity check using greedy decoding
`python ./scripts/generate/test_generate.py --config "./torchtitan/models/llama3/train_configs/llama3_8b_debug.toml" --checkpoint ./outputs/checkpoint-test/step-1 --top_k 1 --max_new_tokens 200 --prompt "Once upon a time in a land far, far away"`
The correct output should be exactly: Screenshot 2025-07-29 at 11 14 15 AM
3. Thorough test of entire model weights using KL divergence on logits after forward passes.
`python ./scripts/checkpoint_conversion/compare_tt_and_tt.py`
You will need to create a correct dcp checkpoint to compare to using the `convert_from_hf.py` script which runs on full tensors.
The correct loss should be in a similar order of magnitude as`-1.3229e-13`

### Closed: fixed by https://github.com/pytorch/pytorch/pull/159656

### CODE REVIEW

**Comment by wwwjn** [comment]
Thank you so much for finding this little tiny bug through detailed test!! Here we find 2 issue related to this bug:

1. view() op + full_tensor()  on a DTensor with [_StridedShard(), Shard()] are not correct. If we change the order , full_tensor() + view() op,  the numerical result are different
Minimal reproduce:  [https://github.com/wwwjn/torchtitan/blob/llama3-sd/scripts/run_permute_test.py](https://github.com/wwwjn/torchtitan/blob/llama3-sd/scripts/fsdp_tp_test.py)
2. If the weight in state_dict is [Shard(0), Shard(0)] , the model applied FSDP +TP so the model expected weights are  [_StridedShard(), Shard()] , the loaded weights through set_model_state_dict are wrong.
Minial reproduce:[ https://github.com/wwwjn/torchtitan/blob/llama3-sd/scripts/fsdp_tp_test.py](https://github.com/wwwjn/torchtitan/blob/llama3-sd/scripts/fsdp_tp_test.py)

We have experts working on it now :)
